sled-agent/src/http_entrypoints.rs

```diff
@@ -64,9 +64,28 @@ async fn services_put(
     rqctx: RequestContext<SledAgent>,
     body: TypedBody<ServiceEnsureBody>,
 ) -> Result<HttpResponseUpdatedNoContent, HttpError> {
-    let sa = rqctx.context();
+    let sa = rqctx.context().clone();
     let body_args = body.into_inner();
-    sa.services_ensure(body_args).await.map_err(|e| Error::from(e))?;
+
+    // Spawn a task to run `services_ensure`: dropshot/hyper will cancel an
+    // endpoint's task if the call times out or is otherwise cancelled, and this
+    // leads to a specific issue where booting all the service zones takes
+    // longer than the progenitor client default timeout. The request times out,
+    // `services_ensure` gets cancelled, and this leaves zones partially
+    // configured but not added to the list of `existing_zones`. When the next
+    // `PUT /services` call is made, `services_ensure` will eventually try to
+    // bring up the same zone it was interrupted at, leading to configuration
+    // issues. See: oxidecomputer/omicron#3098.
```

Comment:
I'd suggest summarizing this, maybe something like:

```suggestion
    // Spawn a task to run `services_ensure` so that it runs to completion even if
    // this future is cancelled (as might happen if the client abandons the request).
    // Otherwise, unexpected cancellation of this future could result in leaving the in-
    // memory state of this zone invalid.
```

There's a lot of other detail there that's pretty specific to the bug you saw.

Tangentially: what I understood happened (which could well be wrong!) is that the _client_ timed out its request, the socket got closed, the corresponding Future got dropped, and so it became cancelled.  As far as I know, neither hyper nor dropshot has a policy that would cause a request to be timed out.

Reply:
I trimmed it in fb7b9f7d, let me know what you think